### PR TITLE
Accumulate Time Lost from comments sent through LOVE 

### PR DESCRIPTION
**Pull request by Vebop**
Added update_time_loss() to request jira ticket by ID, and updates the existing time_lost value adding what was given in the form to the original value from the ticket.
Called this function from jira_comment() only if 'time_lost' is provided in the request_data. 

Updated test_jira_comment() to patch this new function. 
Added test_update_time_loss.

### CODE REVIEW

File: manager/api/tests/test_jira.py [hunk 1]
```diff
@@ -311,6 +316,31 @@ def test_needed_parameters(self):
...
+        response = requests.Response()
+        response.status_code = 200
+        response.json = lambda: {"customfield_10106": 13.6}
+        mock_jira_client.return_value = response
+
+        mock_jira_patcher = patch("requests.put")
+        mock_jira_client = mock_jira_patcher.start()
+        response = requests.Response()
+        response.status_code = 200
+        mock_jira_client.return_value = response
```
Comment (pothiers) [inline manager/api/tests/test_jira.py:333]: Replace with patch decorators to simplify. No need to create response twice, me thinks.

@patch("requests.get")
@patch("requests.post")
def test_update_time_loss(self, mock_get, mock_post):
        """Test call to update_time_loss and verify field was updated"""
        # patch both requests.get, requests.put
        response = requests.Response()
        response.status_code = 200
        mock_get.return_value = response

        response.json = lambda: {"customfield_10106": 13.6}
        mock_post.return_value = response

**Review by pothiers — APPROVED** [review]

File: manager/api/tests/test_jira.py (hunk 1 above)
Comment (Vebop) [reply manager/api/tests/test_jira.py:333]: I considered that and saw the pattern in this file was to use the decorators for patching environment variables and the implemented way for `requests`. I wasn't familiar with any best practices that designate one way or the other. However I do like this way so I'll update it.

File: manager/manager/utils.py [hunk 2]
```diff
@@ -505,6 +505,57 @@ def jira_ticket(request_data):
     )
 
 
+def update_time_loss(jira_id: int, add_time_loss: float = 0.0) -> Response:
+    """Connect to the Rubin Observatory JIRA Cloud REST API to
+    update a jira ticket's specific field
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:510]: We should specify here that were are updating the specific time_lost field.

File: manager/manager/utils.py [hunk 3]
```diff
@@ -505,6 +505,57 @@ def jira_ticket(request_data):
...
+
+    Params
+    ------
+    jira_id: int
+        Jira ID
+
+    Returns
+    -------
+    Response
+        define response here
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:520]: Can you perhaps describe the Response same as the next case, `jira_comment`? something like:
```suggestion
        The response and status code of the request to the JIRA API
```

File: manager/manager/utils.py [hunk 4]
```diff
@@ -505,6 +505,57 @@ def jira_ticket(request_data):
...
+    headers = {
+        "Authorization": f"Basic {os.environ.get('JIRA_API_TOKEN')}",
+        "content-type": "application/json",
+    }
+    get_url = (
+        f"https://{os.environ.get('JIRA_API_HOSTNAME')}/rest/api/latest/issue/{jira_id}"
+    )
+    response = requests.get(get_url, headers=headers)
+    existent_time_loss = (
+        response.json().get("customfield_10106", 0.0)
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:531]: This might be too much to ask, as I should have done it before :sweat_smile: ... but for the sake of having better maintainable software I'll ask anyways... Could you move this constant: `customfield_10106` to a variable (see begining of the `utils.py` file)? Also, if you could move the other `customfield_` constants it would be awesome :pray:.

File: manager/manager/utils.py [hunk 5]
```diff
@@ -505,6 +505,57 @@ def jira_ticket(request_data):
...
+        if response.status_code == 200
+        else 0.0
+    )
+    jira_payload = {
+        "fields": {
+            "customfield_10106": float(existent_time_loss + add_time_loss),
+        },
+    }
+    put_url = f"https://{os.environ.get('JIRA_API_HOSTNAME')}/rest/api/latest/issue/{jira_id}/"
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:540]: I see the difference between `get_url` and `put_url` is only a `/` at the end for the later one. Is this necessary, I'm asking basically if we could use the same url for both purposes.

File: manager/manager/utils.py [hunk 6]
```diff
@@ -505,6 +505,57 @@ def jira_ticket(request_data):
...
+    response = requests.put(put_url, json=jira_payload, headers=headers)
+
+    if response.status_code == 200 or response.status_code == 204:
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:543]: In which case we would be receiving a status_code = 204?

File: manager/manager/utils.py [hunk 7]
```diff
@@ -505,6 +505,57 @@ def jira_ticket(request_data):
...
+        return Response(
+            {
+                "ack": "Jira field updated",
+                "url": f"https://{os.environ.get('JIRA_API_HOSTNAME')}/browse/{jira_id}",
+            },
+            status=200,
+        )
+    return Response(
+        {
+            "ack": "Jira field could not be updated",
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:553]: We should specify here the time_lost field could not be updated.
```suggestion
            "ack": "Jira time_lost field could not be updated",
```

File: manager/manager/utils.py [hunk 8]
```diff
@@ -547,6 +598,12 @@ def jira_comment(request_data):
     }
     url = f"https://{os.environ.get('JIRA_API_HOSTNAME')}/rest/api/latest/issue/{jira_id}/comment"
     response = requests.post(url, json=jira_payload, headers=headers)
+
+    if "time_lost" in request_data:
+        update_time_loss(
+            jira_id=jira_id, add_time_loss=request_data.get("time_lost", 0.0)
+        )
```
Comment (sebastian-aranda) [inline manager/manager/utils.py:605]: if the time loss update fails we shouldn't proceed with the rest of the process, I think we should return the response from `update_time_loss` when it returns status_code = 400.
```suggestion
    if "time_lost" in request_data:
        response = update_time_loss(
            jira_id=jira_id, add_time_loss=request_data.get("time_lost", 0.0)
        )
        if response.status_code == 400:
            return response
```

File: manager/api/tests/test_jira.py [hunk 9]
```diff
@@ -311,6 +316,27 @@ def test_needed_parameters(self):
 
         mock_jira_patcher.stop()
 
+    @patch("requests.get")
+    @patch("requests.put")
```
Comment (sebastian-aranda) [inline manager/api/tests/test_jira.py:320]: Not an obligation, but for the sake of standardization could you use the format as on the `test_add_comment` function (without fixtures)? Again, not an obligation.

File: manager/api/tests/test_jira.py [hunk 10]
```diff
@@ -319,6 +345,10 @@ def test_add_comment(self):
         response.status_code = 201
         mock_jira_client.return_value = response
 
+        mock_timeloss_patcher = patch("manager.utils.update_time_loss")
+        mock_timeloss_client = mock_timeloss_patcher.start()
+        mock_timeloss_client.return_value = response
```
Comment (sebastian-aranda) [inline manager/api/tests/test_jira.py:350]: I left a comment before in regards of returning a 400 response in the case the `update_time_loss` fails (on the `jira_comment` function). If you are willing to add that suggestion, that should be tested here (ensuring we get the bad response arranged by the `update_time_loss` function, in this case mocked up).

File: manager/api/tests/test_jira.py [hunk 11]
```diff
@@ -311,6 +316,27 @@ def test_needed_parameters(self):
 
         mock_jira_patcher.stop()
 
+    @patch("requests.get")
+    @patch("requests.put")
+    def test_update_time_loss(self, mock_get, mock_put):
```
Comment (sebastian-aranda) [inline manager/api/tests/test_jira.py:321]: Perhaps you could also test when the `update_time_loss` fails and returns a 400 status code, ensuring we return the arranged `ack` (e.g. "Jira time_lost field could not be updated" in case you accept my other suggestion too).

**Review by sebastian-aranda — CHANGES_REQUESTED** [review]
Good job @Vebop ! This looks goog, but I left several comments for the sake of explicity, hope you can take a look :pray:

File: manager/api/tests/test_jira.py (hunk 9 above)
Comment (sebastian-aranda) [reply manager/api/tests/test_jira.py:320]: Sorry, just saw @pothiers review... One sec, I'm looking at the comments.

Comment (sebastian-aranda) [reply manager/api/tests/test_jira.py:320]: ok, so this one of those things that can be or cannot be jaja. I think is up to the developer, but the thing is (IIRC) this would be the first place on this repo that uses the fixtures format. Again, not an obligation, up to you ;). **But**, we should standardize this at some point and add some details to the `README.md` to point into a standard direction.

File: manager/manager/utils.py (hunk 6 above)
Comment (Vebop) [reply manager/manager/utils.py:543]: This is a good question that made me realize the response is only ever 204 for success, not 200. When the update is sent successfully we receive a 204 No Content.

Comment (sebastian-aranda) [reply manager/manager/utils.py:543]: nice, good to know! Just sharing a nice resource here in case you find it usfeul :hand_over_mouth:: https://http.cat/

File: manager/api/tests/test_jira.py (hunk 11 above)
Comment (Vebop) [reply manager/api/tests/test_jira.py]: I learned that when we use requests.Response() as our patch in the unit tests, that this is not the same as rest_framework.response.Response (that we set in utils.py). So when we start to mock one of these, we cannot then check what the 'ack' would have been, since we are patching that object.

**Review by sebastian-aranda — APPROVED** [review]
Looks good! thanks for the great work here @Vebop :muscle:.
